spec/compiler/codegen/arithmetics_spec.cr

```diff
@@ -281,6 +316,29 @@ describe "Code gen: arithmetic primitives" do
         end
       {% end %}
     {% end %}
+
+    {% begin %}
+      it "raises overflow if not a number" do
+        run(%(
+          require "prelude"
+
+          results = 0_u64
+
+          {% for float_type in [Float32, Float64] %}
+            {% for method, path_type in SupportedIntsConversions %}
+              results <<= 1
+              begin
+                {{float_type}}::NAN.{{method}}
+                results |= 1
+              rescue OverflowError
```

Comment:
I'm having a hard time to understand why trying to convert a NaN to an int would be an overflow. It's not out of bounds like an overflow, a NaN just can't be represented as an integer (makes no sense), and reporting an overflow feels confusing.

Ruby raises a `FloatDomainError (NaN)` exception for example.

In plain Crystal I'd merely `raise ArgumentError.new("NaN") if nan?`.

Reply:
Maybe `OverflowError` could be interpreted more generically to also include conversion errors where the value can't be represented in the target type.

But that's probably not a good idea because it would apply to many other conversion errors as well. So yeah, `ArgumentError` seems like a good idea.

Reply:
There is no `OverflowError` in Ruby; instead they call it `RangeError`, and then make `FloatDomainError` a subclass of it. That means Ruby code that catches overflow errors will also catch domain errors:

```ruby
begin
  (0.0 / 0.0).to_i
rescue RangeError
  puts 1 # => 1
end
```

I think we should copy this or simply rename `OverflowError` to `RangeError`.

Reply:
Still thinking about the best way to approach this. For performance reasons (for the compiler) it seems we should mimic `OverflowError`:

```crystal
# :nodoc:
fun __crystal_raise_nan : NoReturn
  raise ArgumentError.new("not-a-number")
end
```

including making it a "well-known function".

Reply:
IMO overflow and conversion failure are the same class of error. I would use the same error class for both. Currently, we have `OverflowError`. Maybe that name doesn't fit very well with conversion failure. But I'd still suggest to use that. We can see to maybe rename `OverflowError` to a better name that also describes conversion failures (`MathError`? `ArithmeticError`?). It seems easier to go with the error type we already have instead of using a different one and then later introducing a new concept to reunite them. Alternatively, we could introduce a new error type as a parent for `OverflowError` (and maybe merger `OverflowError` into that later).

`ArgumentError` doesn't seem to fit for this.